docs/RationaleSeq.md

```diff
@@ -0,0 +1,70 @@
+# Seq(uential) Dialect Rationale
+
+This document describes various design points of the `seq` dialect, why it is
+the way it is, and current status. This follows in the spirit of other [MLIR
+Rationale docs](https://mlir.llvm.org/docs/Rationale/).
+
+## Introduction
+
+Digital logic is generally split into two orthogonal categories:
+combinational and sequential. CIRCT contains a `comb` dialect to model the
+basic combinational operations and the (future) `seq` dialect which is
+discussed here. The intention of the `seq` dialect is to provide a set of
+stateful constructs which can be used to model sequential logic, independent
+of the output method (e.g. SystemVerilog).
+
+## Definitions
+
+For the sake of precision, we use the following definitions:
+
+- **Latch:** A memory element which is only sensitive to the levels of its
+inputs. Has no clock. Example: SR Latch.
+- **Clocked (gated) latch:** A latch wherein the inputs are gated by a clock.
+Transparent the entire time the clock is high. Generally referred to as a
+"latch". Examples: "gated SR latch", "D latch".
+- **Flip-flop:** An edge-sensitive memory element. Captures the input value
+on one or both clock edges. Variants: posedge FF, negedge FF,
+"edge-sensitive" FF (captures the input value on both edges), resettable FF.
+- **Register:** A synchronous, resettable memory element. Can be implemented
+using any of the above "circuit level" elements.
+
+## The register operation
+
+The `seq.reg` op models an abstract notion of a "register", independent of
+its implementation (e.g. latch, D flip-flop). It is intended to be 'lowered'
+to a specific implementation, which may be an `sv.always_ff` block, a device
+primitive instantiation, or even a feedback loop in all `comb` logic. It has
+three operands and one attribute:
+
+- **input**: The value to be captured at 'clock'. Generally called 'q'.
+Accepts any type, results in the same type.
+- **clock**: Capture 'value' on the positive edge of this signal.
+- **reset**: Signal to set the state to 'resetValue'. Optional.
+- **resetValue**: A constant which the state is set to upon reset. Optional.
+
+```mlir
+%d = seq.reg %input, %clk [, %reset [<resetValue>] ] : $type(input)
+```
```

Comment:
I'm not an expert here, but this doesn't seem like a very general model.  We have multiported rams and many other things that exist in the world with special constraints.  Defining a common abstraction for this will be tricky.  I'd recommend nailing down the details of the other ops first.